tests/vanilla/store.test.tsx

```diff
@@ -592,3 +592,45 @@ it('should update derived atom even if dependances changed (#2697)', () => {
   store.set(primitiveAtom, 1)
   expect(onChangeDerived).toHaveBeenCalledTimes(1)
 })
+
+it('should preserve dependencies when reusing continuable promise', async () => {
+  const baseAtom = atom(0)
+  const refreshAtom = atom(0)
+  let promise = Promise.resolve(-1)
+  let inProgress = false
+  const derivedAtom = atom<Promise<number>, [], void>(
+    (get, { setSelf }) => {
+      get(refreshAtom)
+      if (inProgress) {
+        return promise
+      }
+      promise = Promise.resolve().then(() => {
+        const value = get(baseAtom)
+        inProgress = true
+        // causes the derivedAtom to be re-evaluated, which triggers the inProgress bail
+        setSelf()
+        return value
+      })
+      setTimeout(() => {
+        inProgress = false
+      })
```

Comment:
If this is a microtask delay, doesn't it reproduce the issue?
```suggestion
      Promise.resovle().then(() => {
        inProgress = false
      })
```

Reply:
Yes, still reproducible with a microtask delay.